### PR TITLE
mediatek: add Xiaomi Redmi Router AX6000 support

### PR DESCRIPTION
**Pull request by aiamadeus**
Hardware specification:
  SoC: MediaTek MT7986A 4x A53
  Flash: ESMT F50L1G41LB 128 MB
  RAM: K4A4G165WF-BCWE 512 MB
  Ethernet: 4x 10/100/1000 Mbps
  WiFi1: MT7976GN 2.4GHz ax 4x4
  WiFi2: MT7976AN 5GHz ax 4x4
  Button: Mesh, Reset

Flash instructions:
  1. Gain ssh and serial port access (see the link: https://openwrt.org/toh/xiaomi/redmi_ax6000#installation)
  2. Use ssh or serial port to log in to the router, and execute the following command:
     nvram set boot_wait=on
     nvram set flag_boot_rootfs=0
     nvram set flag_last_success=1
     nvram set flag_boot_success=1
     nvram set flag_try_sys1_failed=8
     nvram set flag_try_sys2_failed=8
     nvram commit
  3. Set a static ip on the ethernet interface of your computer (e.g. default: ip 192.168.31.100, gateway 192.168.31.1)
  4. Download the initramfs image, rename it to initramfs.bin, and host it with the tftp server.
  5. Interrupt U-Boot and run these commands:
      setenv mtdparts nmbm0:1024k(bl2),256k(Nvram),256k(Bdata),2048k(factory),2048k(fip),256k(crash),256k(crash_log),112640k(ubi)
      saveenv
      tftpboot initramfs.bin
      bootm
  6. After openwrt boots up, use scp or luci web to upload sysupgrade.bin to upgrade.

Note:
  1. I don't have this router, thanks to @pengchujin for providing a remote environment for debugging.
  2. The led chip (may be a spi) seems to lack driver, since I don't have this router, I'm not sure what that is.
  3. I haven't actually operated unlocking ssh/serial, my friend has done this, and it seems like it would be too long
      to write it into the commit message. Hope someone can help write the unlocking tutorial into the openwrt wiki. 

Checklist:
* [x] spi-nand
* [x] ethernet
* [x] button
* [x] nmbm
* [x] wifi2g
* [x] wifi5g
* [x] mac
* [ ] led

### CODE REVIEW

**Comment by soxrok2212** [comment]
There is some instruction on the Wiki for rooting the device. https://openwrt.org/toh/xiaomi/redmi_ax6000

**Comment by Ansuel** [comment]
> 3\. I haven't actually operated unlocking ssh/serial, my friend has done this, and it seems like it would be too long
>     to write it into the commit message. Hope someone can help write the unlocking tutorial into the openwrt wiki.

ideally we should first have some instruction and at least link them in the commit description

**Comment by soxrok2212** [comment]
Instructions worked great for me!
```
root@OpenWrt:~# uptime
 12:25:11 up 3 min,  load average: 0.23, 0.18, 0.08
root@OpenWrt:~# uname -a
Linux OpenWrt 5.15.67 #0 SMP Mon Sep 12 05:18:31 2022 aarch64 GNU/Linux
root@OpenWrt:~# cat /etc/os-release 
NAME="OpenWrt"
VERSION="SNAPSHOT"
ID="openwrt"
ID_LIKE="lede openwrt"
PRETTY_NAME="OpenWrt SNAPSHOT"
VERSION_ID="snapshot"
HOME_URL="https://openwrt.org/"
BUG_URL="https://bugs.openwrt.org/"
SUPPORT_URL="https://forum.openwrt.org/"
BUILD_ID="r20615-fbf70f4bca"
OPENWRT_BOARD="mediatek/filogic"
OPENWRT_ARCH="aarch64_cortex-a53"
OPENWRT_TAINTS="no-all"
OPENWRT_DEVICE_MANUFACTURER="OpenWrt"
OPENWRT_DEVICE_MANUFACTURER_URL="https://openwrt.org/"
OPENWRT_DEVICE_PRODUCT="Generic"
OPENWRT_DEVICE_REVISION="v0"
OPENWRT_RELEASE="OpenWrt SNAPSHOT r20615-fbf70f4bca"
root@OpenWrt:~# 
```

LED bar stays on and solid blue for me, so definitely not working but I can live with it for now.

**Comment by aiamadeus** [comment]
> There is some instruction on the Wiki for rooting the device. https://openwrt.org/toh/xiaomi/redmi_ax6000#rooting

Great, I'll add this link to the commit message.
> LED bar stays on and solid blue for me, so definitely not working but I can live with it for now.

Can you see which chip the leds are connected to?
The original fdt has a strange node called HM0807A, I don't know what it is.

**Comment by soxrok2212** [comment]
> Can you see which chip the leds are connected to?
> The original fdt has a strange node called HM0807A, I don't know what it is.

As far as physical visibility, nope. The LEDs are connected via this header at the bottom:

**Comment by DragonBluep** [comment]
I remember kernel size in mediatek target is very close to 4 MiB recently, so it's better to enlarge it to 8 MiB. And for NMBM reserved space, 8 MiB is enough.  Refer to bootlog, block 960 start at 0x7800000:
```
NOTICE:  First info table with writecount 0 found in block 960
NOTICE:  Second info table with writecount 0 found in block 963
```

**Comment by aiamadeus** [comment]
> I remember kernel size in mediatek target is very close to 4 MiB recently, so it's better to enlarge it to 8 MiB. And for NMBM reserved space, 8 MiB is enough.

So I added this line `KERNEL_SIZE := 4096k`. Due to stock uboot limitations, the kernel must be included in the ubi partition.
And with nmbm, all partitions except ubi should be remapped, is it appropriate to directly change the size of bmt-remap-range? Sorry I don't know nmbm very well, so I'm a little confused.

**Comment by DragonBluep** [comment]
I think enlarge bmt-remap-range won't hurt before kernel reach to the end of the remap range.
nmbm will use the last 8MiB as a spare tire. When there is a bad block in the bmt-remap-range, the NMBM layer picks a good block from the last 8MiB and creates a "link" that maps to the address of the bad block. Outside the remap range, it seems that mtd skips the bad blocks directly, which may cause u-boot that supports nmbm failed to load the kernel properly.
image

image

In first image u-boot will load part1+part2+part3, and in second image, u-boot load the wrong kernel (part1+random+part2)

Maybe using `pad-to` to control the size of the kernel to a fixed value is a good option. Although there will be some storage waste, but fortunately we still have 100 MiB+ of space

**Comment by DragonBluep** [comment]
@soxrok2212 Can you try plugging in the power first and waiting for about **1~2 second**, then connect the TTL GND to see if there is any output.

**Comment by aiamadeus** [comment]
> Maybe using `pad-to` to control the size of the kernel to a fixed value is a good option. Although there will be some storage waste, but fortunately we still have 100 MiB+ of space

Thank you so much for your suggestion. I tried to ask mtk, and their reply was that since the kernel is already in ubi, there is no need to care about the size of kernel.

**Comment by ptpt52** [comment]
trying to make it possible to flash via ssh on oem system
this patch to gen an `initramfs-factory.ubi` image to flash it to ubi1 mtd partition
```diff
diff --git a/target/linux/mediatek/image/filogic.mk b/target/linux/mediatek/image/filogic.mk
index 75d461b630..5e28321368 100644
--- a/target/linux/mediatek/image/filogic.mk
+++ b/target/linux/mediatek/image/filogic.mk
@@ -38,6 +38,20 @@ define Build/mt7986-gpt
        rm $@.tmp
 endef
 
+define Build/gen-ubi-initramfs
+       sh $(TOPDIR)/scripts/ubinize-image.sh \
+               $(if $(UBOOTENV_IN_UBI),--uboot-env) \
+               --kernel $(KDIR)/tmp/$(KERNEL_INITRAMFS_IMAGE) \
+               $(foreach part,$(UBINIZE_PARTS),--part $(part)) \
+               "$(1).tmp" \
+               -p $(BLOCKSIZE:%k=%KiB) -m $(PAGESIZE) \
+               $(if $(SUBPAGESIZE),-s $(SUBPAGESIZE)) \
+               $(if $(VID_HDR_OFFSET),-O $(VID_HDR_OFFSET)) \
+               $(UBINIZE_OPTS) && \
+       cat "$(1).tmp" >> "$(1)" && rm "$(1).tmp" && \
+       $(CP) "$(1)" $(BIN_DIR)/
+endef
+
 define Device/bananapi_bpi-r3
   DEVICE_VENDOR := Bananapi
   DEVICE_MODEL := BPi-R3
@@ -137,6 +151,9 @@ define Device/xiaomi_redmi-router-ax6000
   PAGESIZE := 2048
   KERNEL_IN_UBI := 1
   KERNEL_SIZE := 4096k
+  KERNEL_INITRAMFS := kernel-bin | lzma | \
+       fit lzma $$(KDIR)/image-$$(firstword $$(DEVICE_DTS)).dtb with-initrd | \
+       gen-ubi-initramfs $(KDIR)/tmp/$$(KERNEL_INITRAMFS_PREFIX)-factory.ubi
   IMAGE/sysupgrade.bin := sysupgrade-tar | append-metadata
 endef
 TARGET_DEVICES += xiaomi_redmi-router-ax6000
```

possible installation method:
1. Gain ssh.
2. login ssh and flash `initramfs-factory.ubi` to `ubi1` mtd partition, via command: `ubiformat  /dev/mtdXX -y -f /tmp/initramfs-factory.ubi`, where `mtdXX` is the `ubi1` partition mtd device path.
3. change nvram env to boot from ubi1
4. boot into openwrt-initramfs and login ssh, do ubootenv setup, and then flash sysupgrade file via sysupgrade command.

**Comment by nasbdh9** [comment]
It is best if stock u-boot can be replaced like Linksys E8450

**Comment by soxrok2212** [comment]
> @soxrok2212 Can you try plugging in the power first and waiting for about **1~2 second**, then connect the TTL GND to see if there is any output.

I was able to tftpboot again and refresh OpenWrt, but the device still refuses to boot it. Think the console was broken last night because of a bad ground pin.

**Comment by aiamadeus** [comment]
> I was able to tftpboot again and refresh OpenWrt, but the device still refuses to boot it. Think the console was broken last night because of a bad ground pin.

Maybe you can try to recover with the official recovery software of Xiaomi Router, or after sale.

**Comment by aiamadeus** [comment]
> Here is how the UBI partition (mtd7) looks...

It seems need to ubiformat and sysupgrade again.

**Comment by soxrok2212** [comment]
Looks like the sys upgrade isn't writing to UBI correctly. Sometimes it loops like the image above, others it writes a bunch of junk. Stuck in boot loop.

**Comment by soxrok2212** [comment]
@aiamadeus could you share the output of `printenv` from u-boot? remove all identifying data of course. trying to see if something is off. I build a fresh image and re-flashed, but it's still stuck in a boot loop.

**Comment by aiamadeus** [comment]
@soxrok2212 Can you try setting uboot env like this?
```
setenv boot_fw1 run boot_rd_img\;bootm
setenv flag_boot_rootfs 1
setenv flag_last_success 1
```
Then the router will boot from ubi or ubi1?

**Comment by soxrok2212** [comment]
> @soxrok2212 Can you try setting uboot env like this?
> 
> ```
> setenv boot_fw1 run boot_rd_img\;bootm
> setenv flag_boot_rootfs 1
> setenv flag_last_success 1
> ```
> 
> Then the router will boot from ubi or ubi1?

This puts it into a boot loop, so yes looks like it tries booting from ubi1. From there, reset `flag_try_sys1_failed` and `flag_try_sys2_failed` both to 0 and it boots normally.

**Comment by aiamadeus** [comment]
> This puts it into a boot loop, so yes looks like it tries booting from ubi1. From there, reset `flag_try_sys1_failed` and `flag_try_sys2_failed` both to 0 and it boots normally.

So if `flag_boot_rootfs=1` even if we change boot_rd_img2, will the router still boot from the ubi1 partition?
It looks like vendor uboot modifies the env by itself, not sure how redmi ax6s avoids this problem.

**Comment by soxrok2212** [comment]
> > This puts it into a boot loop, so yes looks like it tries booting from ubi1. From there, reset `flag_try_sys1_failed` and `flag_try_sys2_failed` both to 0 and it boots normally.
> 
> So if `flag_boot_rootfs=1` even if we change boot_rd_img2, will the router still boot from the ubi1 partition? It looks like vendor uboot modifies the env by itself, not sure how redmi ax6s avoids this problem.

Looks like its booting normally even though both flag_try_sys1_failed and flag_try_sys2_failed counters are not 0. The boot loader algorithm is here, from the forums. https://pastebin.com/xzRXCbzU

**Comment by dangowrt** [comment]
> It is best if stock u-boot can be replaced like Linksys E8450

Should be easily possible, just pick the right combination for storage and DRAM and check ARM Trusted Firmware and U-Boot are done for the Bananapi BPi-R3.

File: package/boot/uboot-envtools/files/mediatek_filogic
Comment (remittor) [inline package/boot/uboot-envtools/files/mediatek_filogic:45]: To support the `fw_setsys` and `fw_printsys` commands, change this line:
`config_foreach ubootenv_add_app_config ubootenv`
to
`config_foreach ubootenv_add_app_config`

**Comment by remittor** [comment]
> * Download the initramfs image, rename it to initramfs.bin, and host it with the tftp server.
> * Interrupt U-Boot and run these commands:
>   setenv mtdparts nmbm0:1024k(bl2),256k(Nvram),256k(Bdata),2048k(factory),2048k(fip),256k(crash),256k(crash_log),112640k(ubi)
>   saveenv
>   tftpboot initramfs.bin
>   bootm

And why is everything so difficult then? Users here can do terrible things!

Here is the simplest solution: https://forum.openwrt.org/t/adding-openwrt-support-for-xiaomi-redmi-router-ax6s-xiaomi-router-ax3200/111085/938

**Comment by soxrok2212** [comment]
> > * Download the initramfs image, rename it to initramfs.bin, and host it with the tftp server.
> > * Interrupt U-Boot and run these commands:
> >   setenv mtdparts nmbm0:1024k(bl2),256k(Nvram),256k(Bdata),2048k(factory),2048k(fip),256k(crash),256k(crash_log),112640k(ubi)
> >   saveenv
> >   tftpboot initramfs.bin
> >   bootm
> 
> And why is it so difficult? Users here can do terrible things!
> 
> Here is the simplest solution: https://forum.openwrt.org/t/adding-openwrt-support-for-xiaomi-redmi-router-ax6s-xiaomi-router-ax3200/111085/938

I imagine for some people they like to see these basic steps that are happening because they know what exactly is happening instead of rather blindly trusting a script. 

That is fantastic for those who don’t know/care though, so appreciate the contribution :)

**Comment by aiamadeus** [comment]
> And why is everything so difficult then? Users here can do terrible things!

You have to ask the manufacturer (Xiaomi), they restrict users to flash the router so much.
The stock uboot of redmi ax6000 refuses to boot when the kernel and ubi are separated,
so we can't flash it like redmi ax6s or other xiaomi routers. 
Maybe the code given by @ptpt52 is a good way, but I can't test if it works.
Btw, using the serial port to flash the firmware is a normal behavior in openwrt.
For example, see: https://github.com/openwrt/openwrt/commit/da8428d277cd3373b05330cb3b4f93aef717c5ab

**Comment by remittor** [comment]
> The stock uboot of redmi ax6000 refuses to boot when the kernel and ubi are separated,

This is pretty weird. The RB06 bootloader should not differ much from the RB03 bootloader.
Can you provide a log in which the bootloader gives an error?
Can you provide a binary image of the bootloader?

**Comment by ptpt52** [comment]
step by step flash openwrt without ttl access

1. Gain ssh and telnet access, see the link below:
     https://openwrt.org/toh/xiaomi/redmi_ax6000#installation

2. ssh/telnet login and check current system is ubi or ubi1
     cat /proc/cmdline
     sample OUTPUT:
```
console=ttyS0,115200n1 loglevel=8 firmware=1 uart_en=1
``` 
if firmware=1, current system is ubi1
if firmware=0, current system is ubi

3. setup nvram:
     if current system is ubi1, please setup: next time it would boot from ubi
```
     nvram set boot_wait=on
     nvram set flag_boot_rootfs=0
     nvram set flag_last_success=0
     nvram set flag_boot_success=1
     nvram set flag_try_sys1_failed=0
     nvram set flag_try_sys2_failed=0
     nvram commit
```
if current system is ubi, please setup: next time it would boot from ubi1
```
     nvram set boot_wait=on
     nvram set flag_boot_rootfs=1
     nvram set flag_last_success=1
     nvram set flag_boot_success=1
     nvram set flag_try_sys1_failed=0
     nvram set flag_try_sys2_failed=0
     nvram commit
```

4. flash initramfs-factory.ubi to system and boot into openwrt initramfs
     if firmware=1, current system is ubi1, flash to ubi
       `ubiformat /dev/mtd8 -y -f /tmp/initramfs-factory.ubi`
     if firmware=0, current system is ubi, flash to ubi1
       `ubiformat /dev/mtd9 -y -f /tmp/initramfs-factory.ubi`

     Note: cat /proc/mtd to lookup the mtd name, here ubi is mtd8 ubi1 is mtd9, please note that is may be diff on your device
```
root@XiaoQiang:~# cat /proc/mtd
dev:    size   erasesize  name
mtd0: 08000000 00020000 "spi0.1"
mtd1: 00100000 00020000 "BL2"
mtd2: 00040000 00020000 "Nvram"
mtd3: 00040000 00020000 "Bdata"
mtd4: 00200000 00020000 "Factory"
mtd5: 00200000 00020000 "FIP"
mtd6: 00040000 00020000 "crash"
mtd7: 00040000 00020000 "crash_log"
mtd8: 01e00000 00020000 "ubi"
mtd9: 01e00000 00020000 "ubi1"
mtd10: 03200000 00020000 "overlay"
```

5. reboot and boot to openwrt initrmfs system, login ssh, then setup nvram/uboot-env again:
```
     fw_setenv flag_boot_rootfs 0
     fw_setenv flag_last_success 0
     fw_setenv mtdparts "nmbm0:1024k(bl2),256k(Nvram),256k(Bdata),2048k(factory),2048k(fip),256k(crash),256k(crash_log),112640k(ubi)"
```

7. flash sysupgrade.bin via sysupgrade

done.

**Comment by remittor** [comment]
@ptpt52 , But wouldn't this method work?
```
nvram set flag_boot_rootfs=0
nvram set flag_last_success=0
nvram set flag_boot_success=1
nvram set flag_try_sys1_failed=0
nvram set flag_try_sys2_failed=0
nvram set mtdparts "nmbm0:1024k(bl2),256k(Nvram),256k(Bdata),2048k(factory),2048k(fip),256k(crash),256k(crash_log),112640k(ubi)"
nvram commit
mtd erase ubi1
mtd -r -e ubi write /tmp/factory.ubi ubi
```
Show me where I'm wrong.

**Comment by ptpt52** [comment]
> @ptpt52 , But wouldn't this method work?
> 
> ```
> nvram set flag_boot_rootfs=0
> nvram set flag_last_success=0
> nvram set flag_boot_success=1
> nvram set flag_try_sys1_failed=0
> nvram set flag_try_sys2_failed=0
> nvram set mtdparts "nmbm0:1024k(bl2),256k(Nvram),256k(Bdata),2048k(factory),2048k(fip),256k(crash),256k(crash_log),112640k(ubi)"
> nvram commit
> mtd erase ubi1
> mtd -r -e ubi write /tmp/factory.ubi ubi
> ```
> 
> Show me where I'm wrong.

1. it is better use `ubiformat`
2. you must first check current system is ubi or ubi1, by `step 2`

**Comment by remittor** [comment]
> 1. it is better use `ubiformat`

I'm sure there are no bad blocks. I will upgrade to a new firmware one day. What is the problem here?

> 2. you must first check current system is ubi or ubi1, by `step 2`

Why is this needed? The mtd driver will overwrite everything and immediately make a hard reboot.

**Comment by ptpt52** [comment]
> This information is used only by the bootloader! After a hard reboot, bootloader will just search for the kernel in the area previously occupied by partitions ubi and ubi1.

no really true.
1. I change nvram vars to switch it to boot from aonther system(ubi or ubi1)
2. I flash initramfs-factory.ubi to ubi1 or ubi, it next boot it boot to the initramfs system so I can flash openwrt sysupgrade.bin

**Comment by ptpt52** [comment]
> This information is used only by the bootloader! After a hard reboot, bootloader will just search for the kernel in the area previously occupied by partitions ubi and ubi1.

no really true.
1. I change nvram vars to switch it to boot from aonther system(ubi or ubi1)
2. I flash initramfs-factory.ubi to ubi1 or ubi, it next boot it boot to the initramfs system so I can flash openwrt sysupgrade.bin

**Comment by ptpt52** [comment]
@remittor 
oops
it looks like your method is ok ? first we have to create a factory.ubi image

**Comment by remittor** [comment]
> first we have to create a factory.ubi image

That's right. I forgot to write about it right away.

**Comment by ptpt52** [comment]
> @remittor oops it looks like your method is ok ? first we have to create a factory.ubi image


to gen a factory.bin
```
diff --git a/target/linux/mediatek/image/filogic.mk b/target/linux/mediatek/image/filogic.mk
index 2c9ddcb86f..3ff32bdac0 100644
--- a/target/linux/mediatek/image/filogic.mk
+++ b/target/linux/mediatek/image/filogic.mk
@@ -150,6 +150,9 @@ define Device/xiaomi_redmi-router-ax6000
   BLOCKSIZE := 128k
   PAGESIZE := 2048
   KERNEL_IN_UBI := 1
+  IMAGE_SIZE := 30720k
+  IMAGES += factory.bin
+  IMAGE/factory.bin := append-ubi | check-size $$$$(IMAGE_SIZE)
   KERNEL_INITRAMFS := kernel-bin | lzma | \
        fit lzma $$(KDIR)/image-$$(firstword $$(DEVICE_DTS)).dtb with-initrd | \
        gen-ubi-initramfs $(KDIR)/tmp/$$(KERNEL_INITRAMFS_PREFIX)-factory.ubi

```

**Comment by ptpt52** [comment]
> @ptpt52 , But wouldn't this method work?
> 
> ```
> nvram set flag_boot_rootfs=0
> nvram set flag_last_success=0
> nvram set flag_boot_success=1
> nvram set flag_try_sys1_failed=0
> nvram set flag_try_sys2_failed=0
> nvram set mtdparts "nmbm0:1024k(bl2),256k(Nvram),256k(Bdata),2048k(factory),2048k(fip),256k(crash),256k(crash_log),112640k(ubi)"
> nvram commit
> mtd erase ubi1
> mtd -r -e ubi write /tmp/factory.ubi ubi
> ```
> 
> Show me where I'm wrong.


reported by somebody, the current system partition is readonly and cannot write to. so this method is not possible

**Comment by ptpt52** [comment]
@dangowrt  hi, is hardware offload supported on mt7986?

**Comment by aiamadeus** [comment]
> is hardware offload supported on mt7986?

Not supported in mainline, see also: https://lore.kernel.org/all/cover.1662661555.git.lorenzo@kernel.org/

**Comment by remittor** [comment]
> reported by somebody, the current system partition is **readonly** and cannot write to. so this method is not possible

reported by somebody:
```
~# cat /sys/block/mtdblock7/ro
0
~# cat /sys/block/mtdblock8/ro
0
```
How else can you check for sure?

**Comment by soxrok2212** [comment]
After setting the following (ensuring to change boot_fw1 so it boots "first image":

```
nvram set boot_wait=on
nvram set flag_boot_rootfs=0
nvram set flag_last_success=1
nvram set flag_boot_success=1
nvram set flag_try_sys1_failed=8
nvram set flag_try_sys2_failed=8
nvram set boot_fw1 run boot_rd_img\;bootm
```

Everything seems to be up and stable. I have this running as an AP forwarding VLAN traffic and it seems to be working fine. I haven't done any formal tests to try to trigger it to break again as it did before (outside of what made it break the first time - unplugging and replugging a few times) but it hasn't seem to have broken since setting boot_fw1 to be the same as boot_fw0.

You've got a thumbs up from me :)

**Review by soxrok2212 — APPROVED** [review]

**Comment by ptpt52** [comment]
> boot_fw1

you don't need to change boot_fw1, since the stock uboot never use it. it is useless.
just `flag_try_sys1_failed >= 6` and `flag_try_sys2_failed >= 6` and `flag_last_success=1` would (force) lead to uboot boot system 0

**Comment by aiamadeus** [comment]
@dangowrt Could you review and merge this? Everything works fine for me now except for the missing driver for the led chip.

**Comment by ptpt52** [comment]
I request to adjust the flash partition to avoid modifying the uboot env of `mtdparts`
```diff
diff --git a/target/linux/mediatek/dts/mt7986a-xiaomi-redmi-router-ax6000.dts b/target/linux/mediatek/dts/mt7986a-xiaomi-redmi-router-ax6000.dts
index a290f5b10c..148f167496 100644
--- a/target/linux/mediatek/dts/mt7986a-xiaomi-redmi-router-ax6000.dts
+++ b/target/linux/mediatek/dts/mt7986a-xiaomi-redmi-router-ax6000.dts
@@ -187,15 +187,20 @@
                                read-only;
                        };
 
+                       partition@600000 {
+                               label = "ubi_kernel";
+                               reg = <0x600000 0x1e00000>;
+                       };
+
                        /* ubi partition is the result of squashing
                         * consecutive stock partitions:
                         * - ubi
                         * - ubi1
                         * - overlay
                         */
-                       partition@600000 {
+                       partition@2400000 {
                                label = "ubi";
-                               reg = <0x600000 0x6e00000>;
+                               reg = <0x2400000 0x5000000>;
                        };
 
                        /* last 12 MiB is reserved for NMBM bad block table */
```

in this way we don't need to change `mtdparts` and keep it as stock value
```
mtdparts=nmbm0:1024k(bl2),256k(Nvram),256k(Bdata),2048k(factory),2048k(fip),256k(crash),256k(crash_log),30720k(ubi),30720k(ubi1),51200k(overlay)
```
this is very helpfull to make it possible to revert back to stock firmware using the Xiaomi Firmware Tools
According to the actual test, if mtdparts is modified, it will break Xiaomi Firmware Tools to work normally

the new partition changes is here https://github.com/x-wrt/x-wrt/commit/6ef69f9d4a05811acb584b0e8736d91e97a91b5c

This implementation is to keep the 30M ubi_kernel partition aligned with the stock ubi partition, and put the kernel volume into the it, the stock uboot could boot from it, without changing the mtdparts

Finally, we have 30+50M as ubi partition to store rootfs and rootfs_data volume into it.

**Comment by soxrok2212** [comment]
For some users reverting would be great. I’ll personally never go back to a fw that calls home though… :)

**Comment by aiamadeus** [comment]
@soxrok2212 Can you test the replacement of u-boot? I'm not sure if this will make it bricks.

**Comment by soxrok2212** [comment]
I’m a little hesitant to do that, especially now that setting flag_try_sys1_failed >= 6 and flag_try_sys2_failed >= 6 seems to have fixed that bricked device.

**Comment by remittor** [comment]
> I'd imagine most users would like to stick with OpenWrt

The ability to easily return to stock firmware should be required! See: https://github.com/openwrt/openwrt/pull/5063

> the stock fw is only available in Chinese if no other reason.

I made a very easy to use stock firmware translator (EN/RU): https://forum.openwrt.org/t/adding-openwrt-support-for-xiaomi-redmi-router-ax6s-xiaomi-router-ax3200/111085/938
Sources: https://github.com/openwrt-xiaomi/xmir-patcher
This method supports all xiaomi routers.

**Comment by dangowrt** [comment]
> > I'd imagine most users would like to stick with OpenWrt
> 
> The ability to easily return to stock firmware should be required! See: #5063

Having the option to return to stock firmware is nice to have, especially while the device is still very new.
However, please consider that in OpenWrt we will most likely offer support for this hardware **much longer** than the vendor firmware will receive any updates. Today there are many devices which are 10+ years old are still actively maintained and supported in OpenWrt, vendors usually update firmware for about 2-3 years before declaring the hardware being EOL.
Having that in mind, I wouldn't accept **any compromise** in functionality when running OpenWrt just for being able to revert to stock -- because anyway this is only relevant now in the beginning of the shelf-life of a device till the vendor declares the firmware EOL, after that nobody will ever want to return to a then outdated stock firmware.

So having 10% less flash space is absolutely ok, of course. But if keeping compatibility with the stock firmware means having to live with more limiting choices of the vendor I'd always be in favor of getting the most out the device when running OpenWrt, because that's what the project is all about in the end.

**Comment by soxrok2212** [comment]
@dangowrt could we get your approval to start 1 of 6? :)

**Comment by EUA** [comment]
> > Can you see which chip the leds are connected to?
> > The original fdt has a strange node called HM0807A, I don't know what it is.
> 
> As far as physical visibility, nope. The LEDs are connected via this header at the bottom: 

4 pin for just a led? I bet that LED is controlled from I2C bus...

File: target/linux/mediatek/dts/mt7986a-xiaomi-redmi-router-ax6000.dts
Comment (aiamadeus) [reply target/linux/mediatek/dts/mt7986a-xiaomi-redmi-router-ax6000.dts]: Thank you, I deleted this bmt-remap-range property.

**Comment by rjocoleman** [comment]
> @soxrok2212 Can you test the replacement of u-boot? I'm not sure if this will make it bricks.

@aiamadeus do you still need some testing on u-boot replacement? (or anything else) I have a couple of these. I'm going to dump the nand shortly and then should be able to do some testing if that would be helpful.

Manufacture date on the box is 2202.07 and came out of the box with 1.0.48 firmware. I've got root and checked that mtd7/mtd8 partitions are read-write.

The LEDs appear to have embedded drivers. I haven't identified which yet but they are on the SPI bus, the 4 pin header goes to a LED board under the heat sink with two packages on it. 
IMG_4125

**Comment by soxrok2212** [comment]
What's left to get this merged in?

**Comment by PussAzuki** [comment]
Most of us can't open this lower cover perfectly, so I think we should give priority to submitting solutions that can be TTL-free?

**Comment by PussAzuki** [comment]
If using LEDs requires proprietary drivers, can we skip the LED-related ones first and then merge this new device in first?

**Comment by soxrok2212** [comment]
Up and stable for about 2 weeks for me. 0 issues so far. Device is rock stable and MT7986 is great :)

**Comment by Ansuel** [comment]
> If using LEDs requires proprietary drivers, can we skip the LED-related ones first and then merge this new device in first?

imho this should be the best solution currently.

**Comment by shauno100** [comment]
I'm a complete newbie when it comes to the code here, I have a Redmi AX6000 myself and am eagerly awaiting OpenWRT support. I can see that 6 approvers are required for this PR to be merged into the OpenWRT master branch. So far there's 1 approval and 2 with review comments so essentially 5 more approvals are needed. The PR seems to be sound code wise according to the comments here so what is the hold up?

**Comment by aiamadeus** [comment]
> The PR seems to be sound code wise according to the comments here so apart from the LEDs which can be revisited later (would majority of users care about this if option for OpenWRT support was granted to begin with?), what is the hold up? The LED strip as far as i am concerned is a waste of time to begin with, all multiple dots of one color with no symbols showing what each LED means, purely aesthetics rather than usefulness in my opinion.

IMHO this has nothing to do with leds. Generally speaking, most pull requests to add Xiaomi routers take longer to merge.
On the one hand, Xiaomi constantly hardens the original system to prevent users from flashing their routers. So we have to spend more and more time to crack the original system, the commit message is getting longer and longer (such as 3343ca7e6837b2ac5f237ea78bf73d50831dea20), and the operation process is getting more and more complicated. This significantly increases the time cost of submission and review.
On the other hand, Xiaomi has never followed the GPL agreement. So that's why we're at a loss with leds right now.
Sorry for the vender-focused remarks, but I think it's much better than questioning the community ;(
Most people (like me) are busy with study and work, and boring exams. So if you really need to flash a router which
is officially supported by openwrt, my personal suggestion is to switch to a router with a simpler flashing process.
Or go to the forums to get tutorials and the firmware pre-compiled by others.

**Comment by stoickish** [comment]
I just got one of these devices yesterday, and I was able to get my device flashed successfully.  My observations thus-far have been:

- I had to resort to UART to get it working.  I tried using SSH using [this patch](https://github.com/openwrt/openwrt/pull/10690#issuecomment-1253291107) to create a 'factory.bin' and [these instructions](https://github.com/openwrt/openwrt/pull/10690#issuecomment-1253184148) and it just put the device into a boot loop
- I could just be stupid, but I couldn't find any information on the UART pinouts.  For anyone who cares, if you're looking at the device from the bottom with the antennas at the top, UART pinouts (from left to right) seemed to be TX, GND, RX, VCC
- Minor nitpick but it seems like the numbering of the ports as they're painted on the device don't line up with how they're labeled in OpenWRT.  Like on the device, the WAN port is marked as "1" and the three LAN ports are 2-4, whereas in OpenWRT the LAN ports are 1-3 and in reverse order (like port '4' on the device correlates with 'lan1' in OpenWRT)
- Not sure if it's a bug in OpenWRT itself or if it has something to do with this specific device, but I did note that if I configure an 802.1q VLAN on one of the LAN ports, and add that VLAN to a bridge, that it seems like wired devices on that VLAN can communicate with each other but NOT with wireless clients on the same bridge.  This is a configuration I've used successfully on many other devices.

**Comment by aiamadeus** [comment]
> * Minor nitpick but it seems like the numbering of the ports as they're painted on the device don't line up with how they're labeled in OpenWRT.  Like on the device, the WAN port is marked as "1" and the three LAN ports are 2-4, whereas in OpenWRT the LAN ports are 1-3 and in reverse order (like port '4' on the device correlates with 'lan1' in OpenWRT)

Can you try the latest code?

**Comment by stoickish** [comment]
> > * Minor nitpick but it seems like the numbering of the ports as they're painted on the device don't line up with how they're labeled in OpenWRT.  Like on the device, the WAN port is marked as "1" and the three LAN ports are 2-4, whereas in OpenWRT the LAN ports are 1-3 and in reverse order (like port '4' on the device correlates with 'lan1' in OpenWRT)
> 
> Can you try the latest code?

Yes, that fixed it. The OpenWRT UI seems to be consistent with what's printed on the device now.  Thanks!

**Comment by dangowrt** [comment]
Picked. Thank you!

**Comment by dangowrt** [comment]
Because the commit has been picked into the main git branch.
https://github.com/openwrt/openwrt/commit/641e4f2f0479997c71ff6c1f1b63fb9b5c67e687

**Pull request closed without merging**
